saw-central/src/SAWCentral/Crucible/MIR/Builtins.hs

```diff
@@ -1554,17 +1580,45 @@ verifyPoststate cc mspec env0 globals ret mdMap =
          return (Crucible.simErrorReasonMsg err, md, obligation)
 
     matchResult opts sc =
-      case (ret, mspec ^. MS.csRetValue) of
-        (Just r, Just expect) ->
-            let md = MS.ConditionMetadata
-                     { MS.conditionLoc = mspec ^. MS.csLoc
-                     , MS.conditionTags = mempty
-                     , MS.conditionType = "return value matching"
-                     , MS.conditionContext = ""
-                     } in
-            matchArg opts sc cc mspec MS.PostState md r expect
-        (Nothing     , Just _ )     -> fail "verifyPoststate: unexpected mir_return specification"
-        _ -> return ()
+      case (mspec ^. MS.csRet, ret, mspec ^. MS.csRetValue) of
+
+        -- Non-unit function:
+        --
+        -- mir_return has *already* checked that the user-specified SetupValue
+        -- has the correct *type* for the function’s return.
+        --
+        -- What mir_return does NOT check is whether the *actual* value produced
+        -- by symbolic execution matches what the spec requested.
+        --
+        -- This branch enforces that semantic obligation: it compares the
+        -- symbolic execution result `r` (a MIRVal) against the spec’s expected
+        -- value `expect` using matchArg, generating the equality constraints
+        -- needed in the post-state.
+        (Just _, Just r, Just expect) ->
+          let md = MS.ConditionMetadata
+                   { MS.conditionLoc     = mspec ^. MS.csLoc
+                   , MS.conditionTags    = mempty
+                   , MS.conditionType    = "return value matching"
+                   , MS.conditionContext = ""
+                   }
+          in
+          matchArg opts sc cc mspec MS.PostState md r expect
+
+        -- Unit-returning function:
+        --
+        -- There is no post-state obligation about the return value itself,
+        -- because unit-returning functions do not produce a meaningful MIR
+        -- return value. ret = Nothing always.
+        --
+        -- This branch exists solely to prevent valid unit cases (implicit or
+        -- explicit mir_return ()) from falling into the failure case below.
+        (Nothing, Nothing, _) ->
+          pure ()
+
+        -- Any other combination indicates a broken invariant in execMIRSetup
+        -- or verifySimulate and should not occur for well-formed specs.
+        _ ->
+          fail "verifyPoststate (MIR): inconsistent return/type information (internal error)"
```

Comment:
Since this is an internal-only error, it would be better to make this `panic` rather than fail. Something like:


```suggestion
          panic "verifyPoststate (MIR)" ["inconsistent return/type information"]
```

(I opted to remove the "`internal error`" part of the message, as that's implicit in the use of `panic`.)